File: acs-client-test/python/test-2/client-sdk-2.py
# Copyright 2025 Accelerated Cloud Storage Corporation. All Rights Reserved.
import os
import time
import random
import statistics
from acs_sdk import ACSClient, Session

def percentile(values, perc):
    """Calculate percentile of a sorted list of values."""
    sorted_values = sorted(values)
    k = (len(sorted_values) - 1) * perc
    f = int(k)
    c = int(k) + 1 if k < len(sorted_values) - 1 else int(k)
    d0 = sorted_values[f] * (c - k)
    d1 = sorted_values[c] * (k - f)
    return d0 + d1

def calculate_metrics(latencies, operation, data_size=0):
    """Calculate and print metrics for operations."""
    if not latencies:
        print(f"No valid latencies for {operation}")
        return
        
    # Sort latencies for percentile calculations
    sorted_latencies = sorted(latencies)
    
    # Calculate basic statistics
    min_latency = sorted_latencies[0]
    avg_latency = sum(latencies) / len(latencies)
    
    # Calculate throughput in ops/sec
    total_time = sum(latencies)
    ops_per_sec = len(latencies) / total_time if total_time > 0 else 0
    
    # Calculate throughput in GB/sec if data_size is provided
    gb_per_sec = 0
    if data_size > 0:
        total_bytes = data_size * len(latencies)
        gb_per_sec = (total_bytes / total_time) / (1024 * 1024 * 1024) if total_time > 0 else 0
    
    # Calculate percentiles
    p90 = percentile(sorted_latencies, 0.90)
    p95 = percentile(sorted_latencies, 0.95)
    p99 = percentile(sorted_latencies, 0.99)
    
    # Convert latencies from seconds to milliseconds
    min_ms = min_latency * 1000
    avg_ms = avg_latency * 1000
    p90_ms = p90 * 1000
    p95_ms = p95 * 1000
    p99_ms = p99 * 1000
    
    print(f"\n{operation} Metrics:")
    print(f"Min Latency: {min_ms:.2f} ms")
    print(f"Average Latency: {avg_ms:.2f} ms")
    print(f"P90 Latency: {p90_ms:.2f} ms")
    print(f"P95 Latency: {p95_ms:.2f} ms")
    print(f"P99 Latency: {p99_ms:.2f} ms")
    print(f"Throughput: {ops_per_sec:.2f} ops/sec")
    if data_size > 0:
        print(f"Throughput: {gb_per_sec:.6f} GB/sec")

def large_object_test():
    """Test operations with a large 10GB object."""
    print("\n===== LARGE OBJECT TEST =====")
    # Initialize client
    session = Session(region="us-east-1")
    client = ACSClient(session)
    
    try:
        # Create test bucket
        bucket_name = f"large-object-test-{int(time.time_ns())}"
        print(f"\nCreating bucket: {bucket_name}")
        client.create_bucket(bucket_name)
        
        try:
            # Generate 10GB of random data
            object_size = 10 * 1024 * 1024 * 1024  # 10GB in bytes
            print(f"\nGenerating {object_size/(1024*1024*1024):.2f}GB of random data...")
            data = os.urandom(object_size)
            key = "large-object"
            
            # Upload large object
            print("\nUploading large object...")
            start = time.time()
            client.put_object(bucket_name, key, data)
            upload_latency = time.time() - start
            calculate_metrics([upload_latency], "Large Object Upload", object_size)
            
            # Read large object
            print("\nReading large object...")
            start = time.time()
            retrieved_data = client.get_object(bucket_name, key)
            download_latency = time.time() - start
            calculate_metrics([download_latency], "Large Object Download", object_size)
            
            # Verify data integrity
            print("\nVerifying data integrity...")
            if len(retrieved_data) != len(data):
                print(f"Data size mismatch! Original: {len(data)} bytes, Retrieved: {len(retrieved_data)} bytes")
            elif retrieved_data != data:
                print("Data content mismatch!")
            else:
                print("Data integrity verified successfully!")
            
            # Delete large object
            print("\nDeleting large object...")
            start = time.time()
            client.delete_object(bucket_name, key)
            delete_latency = time.time() - start
            calculate_metrics([delete_latency], "Large Object Deletion", object_size)
            
        finally:
            # Clean up - delete bucket
            print(f"\nCleaning up bucket: {bucket_name}")
            client.delete_bucket(bucket_name)
            
    finally:
        # Close client connection
        client.close()

def list_operations_test():
    """Test bucket and object listing operations."""
    print("\n===== LIST OPERATIONS TEST =====")
    # Initialize client
    session = Session(region="us-east-1")
    client = ACSClient(session)
    
    try:
        # Part 1: Bucket List Test
        base_bucket_name = f"list-test-{int(time.time_ns())}"
        num_buckets = 100
        bucket_names = []
        
        # Create 100 buckets
        print(f"\nCreating {num_buckets} buckets...")
        bucket_create_latencies = []
        
        for i in range(num_buckets):
            bucket_name = f"{base_bucket_name}-{i}"
            bucket_names.append(bucket_name)
            
            start = time.time()
            client.create_bucket(bucket_name)
            bucket_create_latencies.append(time.time() - start)
        
        calculate_metrics(bucket_create_latencies, "Bucket Creation")
        
        # List all buckets
        print("\nListing all buckets...")
        list_bucket_latencies = []
        
        for _ in range(10):  # Perform list operation 10 times for more reliable metrics
            start = time.time()
            buckets = client.list_buckets()
            list_bucket_latencies.append(time.time() - start)
        
        calculate_metrics(list_bucket_latencies, "Bucket Listing")
        
        # Delete all buckets
        print(f"\nDeleting {num_buckets} buckets...")
        bucket_delete_latencies = []
        
        for bucket_name in bucket_names:
            start = time.time()
            client.delete_bucket(bucket_name)
            bucket_delete_latencies.append(time.time() - start)
        
        calculate_metrics(bucket_delete_latencies, "Bucket Deletion")
        
        # Part 2: Object List Test
        object_test_bucket = f"object-list-test-{int(time.time_ns())}"
        num_objects = 1000
        
        # Create a bucket for object tests
        print(f"\nCreating bucket for object list test: {object_test_bucket}")
        client.create_bucket(object_test_bucket)
        
        try:
            # Create 1000 small objects
            print(f"\nCreating {num_objects} objects of size 1 byte...")
            object_create_latencies = []
            data = b"0"  # 1 byte of data
            
            for i in range(num_objects):
                key = f"small-object-{i}"
                
                start = time.time()
                client.put_object(object_test_bucket, key, data)
                object_create_latencies.append(time.time() - start)
            
            calculate_metrics(object_create_latencies, "Object Creation", 1)
            
            # List all objects in the bucket
            print("\nListing all objects...")
            list_object_latencies = []
            
            for _ in range(10):  # Perform list operation 10 times
                start = time.time()
                objects = client.list_objects(object_test_bucket)
                list_object_latencies.append(time.time() - start)
            
            calculate_metrics(list_object_latencies, "Object Listing")
            
            # Delete all objects
            print(f"\nDeleting {num_objects} objects...")
            object_delete_latencies = []
            
            for i in range(num_objects):
                key = f"small-object-{i}"
                
                start = time.time()
                client.delete_object(object_test_bucket, key)
                object_delete_latencies.append(time.time() - start)
            
            calculate_metrics(object_delete_latencies, "Object Deletion", 1)
            
        finally:
            # Clean up - delete bucket
            print(f"\nCleaning up bucket: {object_test_bucket}")
            client.delete_bucket(object_test_bucket)
            
    finally:
        # Close client connection
        client.close()

def main():
    print("ACS Client SDK Benchmark - Test Suite 2")
    print("======================================")
    
    large_object_test()
    list_operations_test()

if __name__ == "__main__":
    main() 

File: acs-client-test/golang/test-2/client-sdk-2.go
// Copyright 2025 Accelerated Cloud Storage Corporation. All Rights Reserved.
package main

import (
	"context"
	"fmt"
	"math/rand"
	"os"
	"sort"
	"time"

	// Updating the import path to match project structure
	client "github.com/AcceleratedCloudStorage/acs-sdk-go/client"
)

func main() {
	fmt.Println("ACS Client SDK Benchmark - Test Suite 2")
	fmt.Println("======================================")

	// Run large object test
	largeObjectTest()

	// Run list operations test
	listOperationsTest()
}

// largeObjectTest tests operations with a large 10GB object
func largeObjectTest() {
	fmt.Println("\n===== LARGE OBJECT TEST =====")

	// Initialize client
	cli, err := client.NewClient(&client.Session{
		Region: "us-east-1",
	})
	if err != nil {
		fmt.Printf("Failed to create client: %v\n", err)
		os.Exit(1)
	}
	defer cli.Close()

	ctx := context.Background()

	// Create test bucket
	bucketName := fmt.Sprintf("large-object-test-%d", time.Now().UnixNano())
	fmt.Printf("\nCreating bucket: %s\n", bucketName)
	err = cli.CreateBucket(ctx, bucketName)
	if err != nil {
		fmt.Printf("Failed to create bucket: %v\n", err)
		return
	}

	defer func() {
		// Clean up bucket at the end
		fmt.Printf("\nCleaning up bucket: %s\n", bucketName)
		cli.DeleteBucket(ctx, bucketName)
	}()

	// Generate 10GB of random data
	objectSize := int64(10 * 1024 * 1024 * 1024) // 10GB in bytes
	fmt.Printf("\nGenerating %.2fGB of random data...\n", float64(objectSize)/(1024*1024*1024))
	data := make([]byte, objectSize)
	_, err = rand.Read(data)
	if err != nil {
		fmt.Printf("Failed to generate random data: %v\n", err)
		return
	}

	key := "large-object"

	// Upload large object
	fmt.Println("\nUploading large object...")
	startTime := time.Now()
	err = cli.PutObject(ctx, bucketName, key, data)
	uploadLatency := time.Since(startTime)
	if err != nil {
		fmt.Printf("Failed to upload object: %v\n", err)
		return
	}
	calculateMetricsForBenchmark([]time.Duration{uploadLatency}, "Large Object Upload", int(objectSize))

	// Read large object
	fmt.Println("\nReading large object...")
	startTime = time.Now()
	retrievedData, err := cli.GetObject(ctx, bucketName, key)
	downloadLatency := time.Since(startTime)
	if err != nil {
		fmt.Printf("Failed to download object: %v\n", err)
		return
	}
	calculateMetricsForBenchmark([]time.Duration{downloadLatency}, "Large Object Download", int(objectSize))

	// Verify data integrity
	fmt.Println("\nVerifying data integrity...")
	if len(retrievedData) != len(data) {
		fmt.Printf("Data size mismatch! Original: %d bytes, Retrieved: %d bytes\n", len(data), len(retrievedData))
	} else if string(retrievedData) != string(data) {
		fmt.Println("Data content mismatch!")
	} else {
		fmt.Println("Data integrity verified successfully!")
	}

	// Delete large object
	fmt.Println("\nDeleting large object...")
	startTime = time.Now()
	err = cli.DeleteObject(ctx, bucketName, key)
	deleteLatency := time.Since(startTime)
	if err != nil {
		fmt.Printf("Failed to delete object: %v\n", err)
		return
	}
	calculateMetricsForBenchmark([]time.Duration{deleteLatency}, "Large Object Deletion", int(objectSize))
}

// listOperationsTest tests bucket and object listing operations
func listOperationsTest() {
	fmt.Println("\n===== LIST OPERATIONS TEST =====")

	// Initialize client
	cli, err := client.NewClient(&client.Session{
		Region: "us-east-1",
	})
	if err != nil {
		fmt.Printf("Failed to create client: %v\n", err)
		os.Exit(1)
	}
	defer cli.Close()

	ctx := context.Background()

	// Part 1: Bucket List Test
	baseBucketName := fmt.Sprintf("list-test-%d", time.Now().UnixNano())
	numBuckets := 100
	var bucketNames []string

	// Create 100 buckets
	fmt.Printf("\nCreating %d buckets...\n", numBuckets)
	bucketCreateLatencies := make([]time.Duration, numBuckets)

	for i := 0; i < numBuckets; i++ {
		bucketName := fmt.Sprintf("%s-%d", baseBucketName, i)
		bucketNames = append(bucketNames, bucketName)

		startTime := time.Now()
		err := cli.CreateBucket(ctx, bucketName)
		bucketCreateLatencies[i] = time.Since(startTime)

		if err != nil {
			fmt.Printf("Failed to create bucket %s: %v\n", bucketName, err)
			continue
		}
	}

	calculateMetricsForBenchmark(bucketCreateLatencies, "Bucket Creation", 0)

	// List all buckets
	fmt.Printf("\nListing all buckets...\n")
	listBucketLatencies := make([]time.Duration, 10) // Perform 10 times for reliable metrics

	for i := 0; i < 10; i++ {
		startTime := time.Now()
		_, err := cli.ListBuckets(ctx)
		listBucketLatencies[i] = time.Since(startTime)

		if err != nil {
			fmt.Printf("Failed to list buckets: %v\n", err)
			continue
		}
	}

	calculateMetricsForBenchmark(listBucketLatencies, "Bucket Listing", 0)

	// Delete all buckets
	fmt.Printf("\nDeleting %d buckets...\n", numBuckets)
	bucketDeleteLatencies := make([]time.Duration, len(bucketNames))

	for i, bucketName := range bucketNames {
		startTime := time.Now()
		err := cli.DeleteBucket(ctx, bucketName)
		bucketDeleteLatencies[i] = time.Since(startTime)

		if err != nil {
			fmt.Printf("Failed to delete bucket %s: %v\n", bucketName, err)
			continue
		}
	}

	calculateMetricsForBenchmark(bucketDeleteLatencies, "Bucket Deletion", 0)

	// Part 2: Object List Test
	objectTestBucket := fmt.Sprintf("object-list-test-%d", time.Now().UnixNano())
	numObjects := 1000

	// Create a bucket for object tests
	fmt.Printf("\nCreating bucket for object list test: %s\n", objectTestBucket)
	err = cli.CreateBucket(ctx, objectTestBucket)
	if err != nil {
		fmt.Printf("Failed to create bucket: %v\n", err)
		return
	}

	defer func() {
		// Cleanup bucket at the end
		fmt.Printf("\nCleaning up bucket: %s\n", objectTestBucket)
		cli.DeleteBucket(ctx, objectTestBucket)
	}()

	// Create 1000 small objects
	fmt.Printf("\nCreating %d objects of size 1 byte...\n", numObjects)
	objectCreateLatencies := make([]time.Duration, numObjects)
	data := []byte("0") // 1 byte of data

	for i := 0; i < numObjects; i++ {
		key := fmt.Sprintf("small-object-%d", i)

		startTime := time.Now()
		err := cli.PutObject(ctx, objectTestBucket, key, data)
		objectCreateLatencies[i] = time.Since(startTime)

		if err != nil {
			fmt.Printf("Failed to put object: %v\n", err)
			continue
		}
	}

	calculateMetricsForBenchmark(objectCreateLatencies, "Object Creation", 1)

	// List all objects
	fmt.Printf("\nListing all objects...\n")
	listObjectLatencies := make([]time.Duration, 10) // Perform 10 times

	for i := 0; i < 10; i++ {
		startTime := time.Now()
		_, err := cli.ListObjects(ctx, objectTestBucket, nil)
		listObjectLatencies[i] = time.Since(startTime)

		if err != nil {
			fmt.Printf("Failed to list objects: %v\n", err)
			continue
		}
	}

	calculateMetricsForBenchmark(listObjectLatencies, "Object Listing", 0)

	// Delete all objects
	fmt.Printf("\nDeleting %d objects...\n", numObjects)
	objectDeleteLatencies := make([]time.Duration, numObjects)

	for i := 0; i < numObjects; i++ {
		key := fmt.Sprintf("small-object-%d", i)

		startTime := time.Now()
		err := cli.DeleteObject(ctx, objectTestBucket, key)
		objectDeleteLatencies[i] = time.Since(startTime)

		if err != nil {
			fmt.Printf("Failed to delete object: %v\n", err)
			continue
		}
	}

	calculateMetricsForBenchmark(objectDeleteLatencies, "Object Deletion", 1)
}

func calculateMetricsForBenchmark(latencies []time.Duration, operation string, dataSize int) {
	if len(latencies) == 0 {
		fmt.Printf("No valid latencies for %s\n", operation)
		return
	}

	// Sort latencies for percentile calculations
	sort.Slice(latencies, func(i, j int) bool {
		return latencies[i] < latencies[j]
	})

	// Calculate min and average
	minLatency := latencies[0]
	var totalDuration time.Duration
	for _, latency := range latencies {
		totalDuration += latency
	}
	avgLatency := totalDuration / time.Duration(len(latencies))

	// Calculate throughput in ops/sec
	opsPerSec := float64(len(latencies)) / totalDuration.Seconds()

	// Calculate throughput in GB/sec if data_size is provided
	var gbPerSec float64
	if dataSize > 0 {
		totalBytes := dataSize * len(latencies)
		gbPerSec = (float64(totalBytes) / totalDuration.Seconds()) / (1024 * 1024 * 1024)
	}

	// Calculate percentiles
	p90Index := int(float64(len(latencies)) * 0.9)
	p95Index := int(float64(len(latencies)) * 0.95)
	p99Index := int(float64(len(latencies)) * 0.99)

	p90Latency := latencies[p90Index]
	p95Latency := latencies[p95Index]
	p99Latency := latencies[p99Index]

	// Convert durations to milliseconds for display
	minMs := float64(minLatency.Nanoseconds()) / 1e6
	avgMs := float64(avgLatency.Nanoseconds()) / 1e6
	p90Ms := float64(p90Latency.Nanoseconds()) / 1e6
	p95Ms := float64(p95Latency.Nanoseconds()) / 1e6
	p99Ms := float64(p99Latency.Nanoseconds()) / 1e6

	fmt.Printf("\n%s Metrics:\n", operation)
	fmt.Printf("Min Latency: %.2f ms\n", minMs)
	fmt.Printf("Average Latency: %.2f ms\n", avgMs)
	fmt.Printf("P90 Latency: %.2f ms\n", p90Ms)
	fmt.Printf("P95 Latency: %.2f ms\n", p95Ms)
	fmt.Printf("P99 Latency: %.2f ms\n", p99Ms)
	fmt.Printf("Throughput: %.2f ops/sec\n", opsPerSec)
	if dataSize > 0 {
		fmt.Printf("Throughput: %.6f GB/sec\n", gbPerSec)
	}
}
